### PR TITLE
fix: sponsored tx fee

### LINKED ISSUE #1954: Set fee to zero when using sponsored transactions
Hey everyone 😄 

I've been playing with sponsored transactions, and for the most part everything already works really well from a user experience standpoint. However, the API doesn't handle fees for sponsored transactions very well, and it can end up with users having a negative balance!

Discussed more here: https://github.com/hirosystems/stacks-blockchain-api/issues/857

For sponsored transactions, there are two "fees" - the regular spending condition, and the sponsor spending condition. They each have a "fee", but only the sponsor's spending condition is what has any effect.

Overall, I think it would be great if the wallet simply hard-coded the fee to zero for sponsored transactions. Right now I can't set a zero-fee by any means (i.e. in transaction request or by changing it in the UI), so I'd imagine this would lead to a lot of confusion for end users.

❤️

### LINKED ISSUE #2141: insufficient balance error while signing a sponsored tx
I have a new feature in lnswap.org where a user can initiate a Bitcoin transfer and the swap operator will lock funds in a contract and then sponsor the claim contract call so the user can receive the funds from the contract in a trustless way.

While testing with a wallet with 0 STX balance I get the below error. Why is this check performed? I just need the wallet to sign the tx and not broadcast it but pass it back to the application which will forward the rawTx to backend via API.

image

Here's the frontend code where setSignedTx(swapResponse.id, serializedTx) passes the signed tx to sponsoring party.

```
  const txOptions = {
    contractAddress: contractAddress,
    contractName: contractName,
    functionName: 'claimStx',
    functionArgs: functionArgs,
    network: activeNetwork,
    postConditionMode: PostConditionMode.Deny,
    postConditions,
    sponsored: true,
    sponsorSpendingCondition: {},
    onFinish: data => {
      console.log('Stacks sign onFinish:', data);
      const serializedTx = data.stacksTransaction.serialize().toString('hex');
      setSignedTx(swapResponse.id, serializedTx);
    },
    onCancel: data => {
      console.log('Stacks sign onCancel:', data);
    },
```

Note that since this is a mocknet gitpod environment the contract call fee estimation is failing as well so it may or may not be related.

### PR DESCRIPTION
**Pull request by fbwoolf**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/1742170500).

This PR fixes the sponsored tx fee by setting it to 0. In addition, I have added a button to the test-app so we can test sponsored transactions.

Screen Shot 2022-01-22 at 9 30 31 AM

Screen Shot 2022-01-23 at 11 52 56 AM

cc/ @kyranjamie @beguene @He1DAr

### CODE REVIEW

File: src/app/pages/sign-transaction/sign-transaction.tsx [hunk 1]
```diff
@@ -92,9 +94,13 @@ function SignTransactionBase(): JSX.Element | null {
...
-        validationSchema={yup.object({
-          fee: feeSchema(),
-        })}
+        validationSchema={
+          !isSponsored
+            ? yup.object({
+                fee: feeSchema(),
+              })
+            : null
+        }
```
Comment (kyranjamie) [inline src/app/pages/sign-transaction/sign-transaction.tsx]: Would recommend moving this out of the JSX inline definition now it's conditional

File: test-app/src/components/debugger.tsx [hunk 2]
```diff
@@ -51,6 +55,18 @@ export const Debugger = () => {
...
+  // If need to add more test tokens: STW7PFH79HW1C9Z0SXBP5PTPHKZZ58KK9WP1MZZA
+  const handleSponsoredTransactionBroadcast = async (tx: StacksTransaction) => {
+    const sponsorOptions = {
+      transaction: tx,
+      sponsorPrivateKey: 'b8c6aaef4b6de5e62648a59fff13e389856dff5e58163e676c2cfdf9dd5dd11101',
+      fee: new BN(388),
+    };
+
+    const sponsoredTx = await sponsorTransaction(sponsorOptions);
+    broadcastTransaction(sponsoredTx, network);
```
Comment (kyranjamie) [inline test-app/src/components/debugger.tsx]:
```suggestion
    return broadcastTransaction(sponsoredTx, network);
```
?

File: src/app/components/fee-row/components/transaction-fee.tsx [hunk 3]
```diff
@@ -8,7 +8,7 @@ export function TransactionFee(props: TransactionFeeProps): JSX.Element | null {
   const { fee, isSponsored } = props;
   return (
     <span data-testid={TransactionSigningSelectors.FeeToBePaidLabel}>
-      {isSponsored ? '🎉 sponsored' : fee} STX
+      {isSponsored ? '(sponsored) 0' : fee} STX
```
Comment (kyranjamie) [inline src/app/components/fee-row/components/transaction-fee.tsx]: Not a fan of brackets. Could we change the whole row to something else? 

Not sure what, exactly, but something that clearly indicates its free.

File: src/app/components/fee-row/fee-row.tsx [hunk 4]
```diff
@@ -43,6 +42,10 @@ export function FeeRow(props: FeeRowProps): JSX.Element {
     if (!input.value && !isCustom && feeEstimations.length) {
       helpers.setValue(microStxToStx(feeEstimations[1].fee).toNumber());
     }
+    // Set fee to 0 if the tx is sponsored
```
Comment (kyranjamie) [inline src/app/components/fee-row/fee-row.tsx]: The code reads `if sponsored, set value 0`, imo this comment adds little.

If it isn't clear what is being set to 0 (the fee), rename helpers to `feeHelpers`

**Review by kyranjamie — APPROVED** [review]

File: src/app/components/fee-row/components/transaction-fee.tsx (hunk 3 above)
Comment (fbwoolf) [reply src/app/components/fee-row/components/transaction-fee.tsx]: Yep, wasn't too sure so threw this out there and was hoping for feedback. I know people didn't love the emoji so was attempting to remove it. Open to suggestions from @eugeniadigon or anyone with an better idea.

File: test-app/src/components/debugger.tsx (hunk 2 above)
Comment (fbwoolf) [reply test-app/src/components/debugger.tsx]: This did work but seems like it should be there, thx.

File: src/app/components/fee-row/fee-row.tsx (hunk 4 above)
Comment (fbwoolf) [reply src/app/components/fee-row/fee-row.tsx]: Ok, thx, I'll remove it. 👍

**Review by beguene — APPROVED** [review]

File: src/app/components/fee-row/components/transaction-fee.tsx (hunk 3 above)
Comment (kyranjamie) [reply src/app/components/fee-row/components/transaction-fee.tsx]: In lieu of any better ideas, my suggestion would be to remove the fee row altogether, and in its place add a message that clearly explains "This transaction is sponsored, so no fee is charged", in a box like this, but with the blue/info colour as it isn't a warning. 

image

**Comment by ginny-d** [comment]
A screen with the suggestions mentioned:
sponsored-transaction

We could even go more low-key and just have it as a tooltip. Depending on how important is to give an explanation


sponsored-transaction-1

**Comment by fbwoolf** [comment]
@eugeniadigon fantastic, thanks for the feedback! 👍

**Comment by fbwoolf** [comment]
Implemented:

Screen Shot 2022-01-24 at 12 02 29 PM

File: src/app/common/transactions/broadcast-transaction.ts [hunk 5]
```diff
@@ -25,7 +25,7 @@ export async function broadcastTransaction(options: BroadcastTransactionOptions)
     const isValidTxId = validateTxId(response.txid);
     if (isValidTxId)
       return {
-        txId: response,
+        txId: response.txid,
```
Comment (fbwoolf) [inline src/app/common/transactions/broadcast-transaction.ts:28]: @kyranjamie can you confirm this change is ok? Things were broken in the test-app bc it was expecting a txId as a string. I am thinking maybe you meant it to be just the `txid`?

Comment (kyranjamie) [reply src/app/common/transactions/broadcast-transaction.ts:28]: Yeah, this makes sense. This looks like another issue caused by the stacks/tx update.

**Comment by pseudozach** [comment]
Any idea what release this will go into and when that will happen?
thanks,

**Comment by fbwoolf** [comment]
> Any idea what release this will go into and when that will happen? thanks,

It will be in our next release which is currently going thru QA. We hope to release tomorrow.

**Comment by Eshwari007** [comment]
@fbwoolf Testing the test app locally.
When I connect to a test account on testnet, receive a blank wallet. I test the test app (locally ) with another PR and it's working fine , just that the sponsored tx was not visible as it was diff build.
Screen Shot 2022-02-01 at 10 36 27 PM

**Pull request merged**